packages/react-aria-components/src/DropZone.tsx

```diff
@@ -0,0 +1,160 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {ButtonContext} from './Button';
+import {ContextValue, Provider, RenderProps, SlotProps, useContextProps, useRenderProps} from './utils';
+import {DropOptions, mergeProps, useButton, useClipboard, useDrop, useFocusRing, useHover, useId, usePress, VisuallyHidden} from 'react-aria';
+import {InputContext} from './Input';
+import {isVirtualDragging} from '@react-aria/dnd';
+import {LinkContext} from './Link';
+import React, {createContext, ForwardedRef, forwardRef, useMemo, useRef} from 'react';
+import {TextContext} from './Text';
+import {useHasChild} from '@react-spectrum/utils';
+import {useLabels} from '@react-aria/utils';
+
+export interface DropZoneRenderProps {
+  /**
+   * Whether the dropzone is currently hovered with a mouse.
+   * @selector [data-hovered]
+   */
+  isHovered: boolean,
+  /**
+   * Whether the dropzone is focused, either via a mouse or keyboard.
+   * @selector :focus
+   */
+  isFocused: boolean,
+  /**
+   * Whether the dropzone is keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean,
+  /**
+   * Whether the dropzone is the drop target.
+   * @selector [data-drop-target]
+   */
+  isDropTarget: boolean
+}
+// note: possibly add isDisabled prop in the future
+export interface DropZoneProps extends Omit<DropOptions, 'getDropOperationForPoint' | 'onInsert' | 'onRootDrop' | 'onItemDrop' | 'onReorder'>, RenderProps<DropZoneRenderProps>, SlotProps {}
+
+export const DropZoneContext = createContext<ContextValue<DropZoneProps, HTMLDivElement>>(null);
+
+function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
+  [props, ref] = useContextProps(props, ref, DropZoneContext);
+  let {dropProps, isDropTarget} = useDrop({...props, ref});
+  let {hoverProps, isHovered} = useHover({});
+  let {focusProps, isFocused, isFocusVisible} = useFocusRing({within: true});
+
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let inputRef = useRef<HTMLInputElement>(null);
+  let uploadLinkRef = useRef<HTMLAnchorElement>(null);
+  let uploadButtonRef = useRef<HTMLButtonElement>(null);
+  let hasButton = useHasChild('button[slot=file]', ref);
+  let hasLink = useHasChild('span[slot=file]', ref);
+
+  let hasInput = useHasChild('input[type=file]', ref);
+  let textId = useId();
+  let isVirtualDrag = isVirtualDragging();
+  let labelText = useMemo(() => {
+    if (!isVirtualDrag && hasInput) {
+      return 'Press enter to select a file';
```

Comment:
This label is pretty desktop specific, should we add a check to see if it is a mobile device and have a screen reader specific label for "double tap to select a file" instead? Additionally, it looks like a mobile (Android) screenreader user can focus the DropZone as a whole and double tap that to trigger the file picker instead, will need to double check on iPad. Maybe don't need the visually hidden button if both can focus the dropzone in that case

Reply:
i can confirm that the same behavior occurs on the iPad as well

Reply:
Gotcha, we can wait for further testing and decide if the visually hidden button should only be rendered for non-mobile experiences then